lib/spack/spack/directory_layout.py

```diff
@@ -242,7 +243,7 @@ def write_spec(self, spec, path):
         """Write a spec out to a file."""
         _check_concrete(spec)
         with open(path, 'w') as f:
-            spec.to_yaml(f)
+            spec.to_yaml(f, hash=ht.full_hash)
```

Comment:
This probably merits a comment that despite `{hash}` in the projection being the dag hash, we're writing *all* the metadata.